airflow/providers/databricks/hooks/databricks_sql.py

```diff
@@ -138,15 +141,39 @@ def get_conn(self) -> Connection:
             )
         return self._sql_conn
 
+    @overload
+    def run(
+        self,
+        sql: str | Iterable[str],
+        autocommit: bool = False,
+        parameters: Iterable | Mapping | None = None,
+        handler: None = None,
+        split_statements: bool = True,
+        return_last: bool = True,
+    ) -> None:
+        ...
+
+    @overload
+    def run(
+        self,
+        sql: str | Iterable[str],
+        autocommit: bool = False,
+        parameters: Iterable | Mapping | None = None,
```

Comment:
I think the mapping case must have str keys, so `Mapping[str, Any]`

Reply:
Good note!

Reply:
Looking at the SQLAlchemy source code, we can also probably update `Iterable` to `Iterable[Mapping[str, Any]]`. https://github.com/sqlalchemy/sqlalchemy/blob/f20f45fae24a4818dc8be5d31a46695b9d16c897/lib/sqlalchemy/engine/interfaces.py#L226-L231C1 See here for the type annotation used by the PEP 249 cursor's execute method: https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection.execute

Reply:
Ah OK. I was not sure what the iterable case is for; if it’s for expansion (`executemany`) that’s reasonable.

Reply:
Actually, I'm going to be careful here because Airflow supports SQLAlchemy 1.4, which doesn't have type annotations. I see for example that Pandas does this type annotation for its "params":

```python
    params: list[Any] | Mapping[str, Any] | None = None,
```

with this comment in the docstring:

```
    params : list, tuple or mapping, optional, default: None
        List of parameters to pass to execute method.  The syntax used
        to pass parameters is database driver dependent. Check your
        database driver documentation for which of the five syntax styles,
        described in PEP 249's paramstyle, is supported.
        Eg. for psycopg2, uses %(name)s so use params={'name' : 'value'}.
```

So out of abundance of caution I'll just do `Iterable | Mapping[str, Any]`.

But when Airflow locks to SQLAlchemy 2.x, I think it'll be safe to do `Iterable[Mapping[str, Any]]`.